Source/JavaScriptCore/wasm/WasmOperationsInlines.h

```diff
@@ -98,6 +98,75 @@ inline EncodedJSValue arrayNew(Instance* instance, uint32_t typeIndex, uint32_t
     return JSValue::encode(JSValue(array));
 }
 
+template<typename T>
+EncodedJSValue createArrayFromDataSegment(Instance* instance, FieldType elementType, size_t arraySize,
+    unsigned dataSegmentIndex, unsigned offset, FixedVector<T>&& tempValues) {
+    JSWebAssemblyInstance* jsInstance = instance->owner();
+    JSGlobalObject* globalObject = jsInstance->globalObject();
+
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    // Determine the array length in bytes from the element type and desired array size
+    size_t elementSize = elementType.type.elementSize();
+    uint32_t arrayLengthInBytes = arraySize * elementSize;
+
+    // Copy the data from the segment into the temp `values` vector
+    if (!instance->copyDataSegment(dataSegmentIndex, offset, arrayLengthInBytes, tempValues)) {
+        throwTypeError(globalObject, scope, "Offset plus array size is out of bounds for data segment"_s);
```

Comment:
Is it possible to add a test for this case?

Reply:
I realized that this case actually shouldn't be there, because if this code is being called from B3 or Air, the bounds check was already done and the exception was thrown if necessary. I reorganized the code so that this just returns false without throwing the error, and added a check to the `call_builtin` case for `ArrayNewData` instead, so that the caller throws the exception whether the operation gets called from B3, Air, or LLInt. Let me know what you think.